linkerd/core/src/main/scala/io/buoyant/linkerd/RequestEvaluator.scala

```diff
@@ -0,0 +1,168 @@
+package io.buoyant.linkerd
+
+import com.twitter.finagle._
+import com.twitter.finagle.client.Transporter.EndpointAddr
+import com.twitter.finagle.http._
+import com.twitter.finagle.naming.buoyant.DstBindingFactory
+import com.twitter.util._
+import io.buoyant.namer.DelegateTree._
+import io.buoyant.namer.{DelegateTree, Delegator}
+import io.buoyant.router.RouterLabel
+import io.buoyant.router.RoutingFactory
+import io.buoyant.router.RoutingFactory.BaseDtab
+import io.buoyant.router.context.{DstBoundCtx, DstPathCtx}
+
+class RequestEvaluator(
+  endpoint: EndpointAddr,
+  namer: DstBindingFactory.Namer,
+  dtab: BaseDtab,
+  label: String
+) extends SimpleFilter[Request, Response] {
+
+  private[this] val RequestTracerMaxDepthHeader = "l5d-max-depth"
+
+  private[this] def printEvaluatedRequest(
+    prependText: String,
+    routerLabel: String,
+    serviceName: String,
+    clientName: String,
+    selectedAddress: String,
+    addresses: Option[Set[String]],
+    dtabResolution: List[String]
+  ) = {
+    Seq(
+      prependText,
+      s"""
+         |--- Router: $routerLabel ---
+         |service name: $serviceName
+         |client name: $clientName
+         |selected address: $selectedAddress
+         |addresses: [${addresses.getOrElse(Set.empty).mkString(", ")}]
+         |dtab resolution:
+         |${dtabResolution.map("  " + _).mkString("\n")}
+    """.stripMargin
+    ).mkString("\n").trim
+  }
+
+  private[this] def formatDTree(
+    dTree: DelegateTree[Name.Bound],
+    tree: List[String],
+    searchPath: Path
+  ): List[String] = {
+    dTree match {
+      case Leaf(path, _, bound) if bound.id == searchPath =>
+        tree :+ path.show
+      case Transformation(_, _, value, remainingTree) =>
+        formatDTree(remainingTree, tree :+ value.path.show, searchPath)
+      case Delegate(path, _, remainingTree) =>
+        formatDTree(remainingTree, tree :+ path.show, searchPath)
+      case Union(path, _, remainingWeightedTrees@_*) =>
+        remainingWeightedTrees.map { wd =>
+          formatDTree(wd.tree, tree :+ s"${wd.weight} * ${path.show}", searchPath)
+        }.filter(!_.isEmpty).head
```

Comment:
protip:

`head` is unsafe because it fails on empty lists.  prefer `headOption`:

`.filter(!_.isEmpty).headOption.toList.flatten`

filter+headOption can be combined into `find`:

`.find(!_.isEmpty).toList.flatten`